### PR TITLE
Fetch signatures from galaxy after the dependency resolver runs

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fetch signatures for type 'galaxy' candidates after the dependency resolver runs so it runs only for the final matches

`find_matches` has access to the corresponding requirements for a match, so it's populating the user-requested signatures. The outer scope recreates the Candidate with the complete list before validating+installing.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ansible-galaxy

### CODE REVIEW

File: lib/ansible/galaxy/collection/__init__.py [hunk 1]
```diff
@@ -743,7 +743,8 @@ def install_collections(
 
     keyring_exists = artifacts_manager.keyring is not None
     with _display_progress("Starting collection install process"):
-        for fqcn, concrete_coll_pin in dependency_map.items():
+        for fqcn in dependency_map.keys():
+            concrete_coll_pin = dependency_map[fqcn]
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/__init__.py]: Why is this change needed?

Comment (s-hertel) [reply lib/ansible/galaxy/collection/__init__.py]: To avoid mutating the dictionary while iterating over it.

File: lib/ansible/galaxy/collection/__init__.py [hunk 2]
```diff
@@ -769,6 +770,19 @@ def install_collections(
...
+                    concrete_coll_pin.namespace, concrete_coll_pin.name, concrete_coll_pin.ver
+                )
+
+                concrete_coll_pin = Candidate(
+                    concrete_coll_pin.fqcn,
+                    concrete_coll_pin.ver,
+                    concrete_coll_pin.src,
+                    concrete_coll_pin.type,
+                    frozenset([*concrete_coll_pin.signatures, *signatures])
+                )
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/__init__.py]: This calls for implementing a conversion helper method in the candidate. Smth like this:
```suggestion
                concrete_coll_pin = concrete_coll_pin.with_signatures(*concrete_coll_pin.signatures, *signatures)
```

File: lib/ansible/galaxy/collection/__init__.py (hunk 1 above)
Comment (s-hertel) [reply lib/ansible/galaxy/collection/__init__.py]: I guess that's not actually mutating the dict.

File: lib/ansible/galaxy/collection/__init__.py [hunk 3]
```diff
@@ -769,6 +769,19 @@ def install_collections(
                     "Skipping signature verification."
                 )
 
+            if concrete_coll_pin.type == 'galaxy':
+                signatures = concrete_coll_pin.src.get_collection_signatures(
+                    concrete_coll_pin.namespace, concrete_coll_pin.name, concrete_coll_pin.ver
+                )
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/__init__.py]: Looking at this, I think this also belongs in the `ansible.galaxy.dependency_resolution.dataclasses` module since it essentially calling internals.

Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py]: Could even be something like `concrete_coll_pin = concrete_coll_pin.with_signatures_repopulated()` that could encapsulate calling `get_collection_signatures()`, followed by the suggested below `with_signatures()`...

File: lib/ansible/galaxy/collection/__init__.py [hunk 4]
```diff
@@ -769,6 +769,19 @@ def install_collections(
                     "Skipping signature verification."
                 )
 
+            if concrete_coll_pin.type == 'galaxy':
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/__init__.py:772]: Is this only useful during installation? It seems, such a conversion might belong to `_resolve_depenency_map()` or even deeper, somewher here: https://github.com/ansible/ansible/blob/9faf944/lib/ansible/galaxy/dependency_resolution/providers.py#L376.

If the signatures are taken into consideration by the depresolver itself, wouldn't it cut the undesired backtracking path early?

Comment (s-hertel) [reply lib/ansible/galaxy/collection/__init__.py:772]: Yes, just installation. `_resolve_depenency_map` is used by download too and it's not needed there.

Maybe I misunderstand  what you mean, but signatures are not taken into account by the dependency resolver (besides compiling signatures - all of them on devel, and now on this branch just those which are user-provided and don't add an API call) and would basically prevent this bugfix (unless we do something like overriding `CollectionDependencyResolver.resolve()` anyway to handle it after an initial find_matches for versions specifically).

We could consider signature validity + versions in the dependency resolver (a feature?), but backtracking if signature validation fails seems like it could be problematic (a change in behavior at least) - if someone prefers to install collections from AH but fall back to Galaxy for collections missing from AH; then any collections on AH that fail signature validation may not give an indicator something was wrong if an unsigned version of the collection was found on Galaxy.

Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py:772]: > Yes, just installation. `_resolve_depenency_map` is used by download too and it's not needed there.

I'd argue that when, verification is enabled, download shouldn't place files on disk either. Hence, I'd expect it to be needed there... Perhaps, this is out of the scope of this specific PR, but in general, it seems like a good idea.

> Maybe I misunderstand what you mean, but signatures are not taken into account by the dependency resolver (besides compiling signatures - all of them on devel, and now on this branch just those which are user-provided and don't add an API call) and would basically prevent this bugfix (unless we do something like overriding `CollectionDependencyResolver.resolve()` anyway to handle it after an initial find_matches for versions specifically).

:thinking: I think, taking them into account would be helpful. Although, I'm not sure what you mean by the bugfix prevention here.

> We could consider signature validity + versions in the dependency resolver (a feature?), but backtracking if signature validation fails seems like it could be problematic (a change in behavior at least) - if someone prefers to install collections from AH but fall back to Galaxy for collections missing from AH; then any collections on AH that fail signature validation may not give an indicator something was wrong if an unsigned version of the collection was found on Galaxy.

Ah.. right. This would likely be considered a feature. Though, I wouldn't expect it to produce a different result, only to abort early.
Though, if we take into account the problem you described, it'd probably require some CLI/config toggles for the behavior and logging.
Still, I'd expect that if one requests signature verification, they'd want the entire dependency tree to be verified. Similar to pip's `--require-hashes` — if it's set for at least one distribution, it makes hashes mandatory for all transitive dependencies of the install session. This is what gave me the idea that the signature checking needs to be baked into the depresolver.

Just think about it:
- Collection A (signed on both AH and Galaxy) depends on a collection B (signed on AH but not signed on Galaxy)
- The latest versions are B==1.0.0 on AH and B==2.0.0 on Galaxy
- B==1.0.0 on AH depends on C==2.0.0 (signed on AH, missing from Galaxy)
- B==2.0.0 on Galaxy depends on D (only published on Galaxy and is compromised); D depends on an older version of C==1.0.0 that is present and signed on AH but has a security vulnerability

Now, when the signatures are not taken into account during the resolution, when installing A, it would go like this:
A -> B==2.0.0 (from Galaxy) -> D -> C==1.0.0 (this opens up possibilities for rollback attacks by attacking the supply chain first)

If the signatures are taken into account, then B==2.0.0 on Galaxy would be rejected and produce a different dependency tree which would arguably be more secure:
A -> B==1.0.0 (from AH) -> C==2.0.0 (a version of C with a vulnerability fixed)

I did some googling and even asked ChatGPT (it agrees with me!). Here's an interesting paper from TUF: https://theupdateframework.io/papers/package-management-security-tr08-02.pdf?raw=true — check out "3.2.5 Depends on Everything" which calls out the signature check on download, for example, or "3.2.6 Unsatisfiable Dependencies". Another one https://theupdateframework.io/papers/attacks-on-package-managers-ccs2008.pdf?raw=true also mentions these attack vectors in the "4.2 Package Signatures" section.

To summarize, now that I've done more research, it seems even more important to get it right...

Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py:772]: I've also asked a security engineering friend who confirmed that it's a good idea to take signatures into account to help prevent security flaws.

Comment (s-hertel) [reply lib/ansible/galaxy/collection/__init__.py:772]: It's good food for thought for the future at least!

I think I understand what you meant in your initial comment now though, so I can answer this more decisively:

> If the signatures are taken into consideration by the depresolver itself, wouldn't it cut the undesired backtracking path early?

No, only if there's a compromised collection, which should be an anomaly. This is fixing the general use case where everything is successfully installed (remember installing from Galaxy == no server signatures, but we query just in case because they might be added). We're making an extra Galaxy API call for every possible match, including those which might be backtracked later or won't be the 'latest'. Consider installing a single collection with no version specified. None of the versions are backtracked, but we fetch Galaxy signatures for every single match, not just the one we install.

Just to be clear, this was my implementation suggestion to avoid reintroducing the issue I'm fixing if you decide to add something like `--require-hashes` as an implicit feature:

* (perhaps) add an alternative to `resolve` that resolves + just returns latest 
* resolve the requirements normally (by version)
* validate the latest matches using signature
* If latest matches pass signature validation, return them
* Otherwise, mark the Resolution state with these new incompatibilities, and go back to step 2 using the new state

I'd be all for it without ^ caveat if you wanted to add it behind a toggle though.

To touch on download:

Signature verification happens when the artifact is extracted currently. This occurs for downloaded collections that are passed to `install`, so at the moment the burden is on the user to include any signatures from Galaxy as supplemental signatures. This was discussed a bit, just out of scope for the initial feature. This could perhaps be improved by inspecting tarfile contents + verifying before adding them to the download destination, or by saving signatures somewhere and including the files as supplemental signatures in the requirements.yml that's populated.

> I did some googling and even asked ChatGPT (it agrees with me!).

Haha, love it! I just wish it was better at being correct than sounding correct :smile: It totally gaslit me the last time I used it for something technical, it had me searching for a CLI option that never existed...

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 5]
```diff
@@ -392,7 +392,6 @@ def _find_matches(self, requirements):
 
             if not unsatisfied:
                 if self._include_signatures:
-                    signatures = src_server.get_collection_signatures(first_req.namespace, first_req.name, version)
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py:395]: Having read your comment at https://github.com/ansible/ansible/pull/80334#discussion_r1153488836, did you mean that this place could be conditional on a CLI toggle (as well as taking signatures into consideration in the future)?

File: lib/ansible/galaxy/collection/__init__.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py:772]: Alright. Let's table the security hardening bit until we can properly discuss having it under a feature toggle...

With that in mind, I think that other restructuring suggestions sound reasonable, right? https://github.com/ansible/ansible/pull/80334/files#r1150826206 / https://github.com/ansible/ansible/pull/80334/files#r1150820606

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 5 above)
Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/providers.py:395]: Yeah, my thought was that if it's behind a toggle (in order to take signatures into account here for backtracking), that would at least prevent a performance regression for people just installing from Galaxy (or servers without any signatures) who won't get any benefit from the overhead.

File: lib/ansible/galaxy/collection/__init__.py (hunk 3 above)
Comment (s-hertel) [reply lib/ansible/galaxy/collection/__init__.py]: Fixed (I think). I added a helper for Candidate. I didn't put it in the base class since the attribute is `signature_sources` rather than `signatures` for Requirements.

Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py]: @s-hertel I think you may've misunderstood one aspect of what I suggested — I didn't mean for it to become a `@classmethod`, just a normal instance method that returns a new instance, initialized with the updated values.

File: lib/ansible/galaxy/collection/__init__.py [hunk 6]
```diff
@@ -769,6 +769,9 @@ def install_collections(
...
+                concrete_coll_pin = concrete_coll_pin.with_signatures_repopulated(concrete_coll_pin)
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/__init__.py]: If this were to be used as a true classmethod, it'd be called as `Candidate.with_repopulated_signatures_extending(concrete_coll_pin)`. What I originally meant was just:
```suggestion
                concrete_coll_pin = concrete_coll_pin.with_signatures_repopulated()
```

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py [hunk 7]
```diff
@@ -584,3 +584,15 @@ def __new__(cls, *args, **kwargs):
...
     def __init__(self, *args, **kwargs):
         super(Candidate, self).__init__()
+
+    @classmethod
+    def with_signatures_repopulated(cls, collection):
+        # type: (t.Type[Candidate], Candidate) -> Candidate
+        if collection.type != 'galaxy':
+            raise ValueError(
+                f"Invalid collection type for {collection}: unable to get signatures from a galaxy server."
+            )
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/dataclasses.py]:
```suggestion
            )

        cls = self.__class__
```

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py [hunk 8]
```diff
@@ -584,3 +584,15 @@ def __new__(cls, *args, **kwargs):
 
     def __init__(self, *args, **kwargs):
         super(Candidate, self).__init__()
+
+    @classmethod
+    def with_signatures_repopulated(cls, collection):
+        # type: (t.Type[Candidate], Candidate) -> Candidate
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/dataclasses.py]:
```suggestion
    def with_signatures_repopulated(self, collection):
        """Populate a new Candidate instance with Galaxy signatures.

        :raises TypeError: If the supplied candidate is not sourced from a Galaxy-like index.
        """
        # type: (Candidate) -> Candidate
```

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py [hunk 9]
```diff
@@ -584,3 +584,15 @@ def __new__(cls, *args, **kwargs):
...
+        if collection.type != 'galaxy':
+            raise ValueError(
+                f"Invalid collection type for {collection}: unable to get signatures from a galaxy server."
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/dataclasses.py]: `TypeError` seems to be better-fitting semantically:
```suggestion
            raise TypeError(
                f"Invalid collection type ({collection.type!s}) for {collection!s}: "
                "unable to get signatures from a galaxy server."
```
(also, it might be useful to log the actual type)

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: OTOH, this guard looks like an entry-contract check for a programming error that shouldn't be triggerable by user input so an `AssertionError` would be a natural fit:

```suggestion
            raise AnsibleAssertionError(
                f"Invalid collection type ({collection.type!s}) for {collection!s}: "
                "unable to get signatures from a galaxy server."
```

**Review by webknjaz — COMMENTED** [review]
@s-hertel this is what I had in mind.

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py (hunk 9 above)
Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: I'm using the `__repr__` to display all attributes of the collection, not just the type so it can be identified easily.

I updated to it an AssertionError since this is just for developers to prevent accidentally calling for types that don't need signatures modified.

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py (hunk 8 above)
Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Added

File: lib/ansible/galaxy/collection/__init__.py (hunk 6 above)
Comment (s-hertel) [reply lib/ansible/galaxy/collection/__init__.py]: Thanks for reviewing again. Fixed this.

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py [hunk 10]
```diff
@@ -584,3 +584,13 @@ def __new__(cls, *args, **kwargs):
 
     def __init__(self, *args, **kwargs):
         super(Candidate, self).__init__()
+
+    def with_signatures_repopulated(self):  # type: (Candidate) -> Candidate
+        """Populate a new Candidate instance with Galaxy signatures.
+        :raises AssertionError: If the supplied candidate is not sourced from a Galaxy-like index.
+        """
+        if self.type != 'galaxy':
+            raise AssertionError(f"Invalid collection type for {self}: unable to get signatures from a galaxy server.")
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/dataclasses.py]: @s-hertel in https://github.com/ansible/ansible/pull/80334#discussion_r1157395171 you implied that you wanted to use `repr()` for `self`, is that correct?
When you inject a variable into an f-string, it uses `str()`: `f'{var}'` becomes `f'{var !s}'` implicitly. To make use of `repr()`, use another specifier — `f'{var !r}'` .
```suggestion
            raise AssertionError(f"Invalid collection type for {self !r}: unable to get signatures from a galaxy server.")
```

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: I also noticed that you didn't use `AnsibleAssertionError` but didn't understand why. https://docs.ansible.com/ansible/latest/dev_guide/testing/sanity/no-assert.html recommends using it whenever possible:
> [..] When possible, utilize a more specific exception than `AnsibleAssertionError`.

Did you just want to avoid importing this exception? What's the motivation here?

```suggestion
            raise AnsibleAssertionError(
                f"Invalid collection type for {self !r}: "
                "unable to get signatures from a galaxy server.",
            )
```

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Per https://github.com/ansible/ansible/blob/c1e19e4bddd13425aba45733372f2a676506256c/lib/ansible/errors/__init__.py#LL237C7-L237C27 `AnsibleInternalError` could be more precise, though...

Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Because I had no idea there was a convention, or that there was documention for it, and I remembered you raising  assertion errors in your dependency resolver PR.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Ah.. Yeah. Bare `AssertionError` is fine, I just remembered that this sanity check recommends an internal one. I'm fine with stdlib personally.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: @s-hertel also, if you didn't intend to use `!r`, feel free to mark this thread as resolved. I don't have a strong preference, just wanted to ask about the inconsistencies I noticed...

**Review by webknjaz — APPROVED** [review]

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py (hunk 10 above)
Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Also, the repr issue was because I forgot we define `__str__` in addition to `__repr__`.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Good thing we have reviews :)

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: This is the reason I prefer explicit specifiers everywhere, so that it's obvious what's expected on the calling side without relying on some disconnected definitions the reader might not even know to look up...

Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: I'm a bit torn over adding it honestly since it's to prevent developer error in the future - would better to have tests... I'd guess there are other places that would be more helpful to use __repr__ than __str__. Why is the latter even defined?

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Well, `repr()` is for devs/REPL/tracebacks/etc. while `str()` is for user-facing output usually.

Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: `__str__` won't necessarily identify the requirement it came from though, I think both devs and users benefit more from the `__repr__` implementation.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Not sure but we could define these things better, in a separate effort. Typically `repr()` would also show `id(self)` somewhere in its output, for example. Just need to double-check that none of the internals rely on `str()`. I can imagine that something in the resolver might.

FWIW I should've added the docstrings for clarity when implementing these initially, but probably forgot. Magic methods deserve them just as much as other public APIs...

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py [hunk 11]
```diff
@@ -584,3 +584,13 @@ def __new__(cls, *args, **kwargs):
 
     def __init__(self, *args, **kwargs):
         super(Candidate, self).__init__()
+
+    def with_signatures_repopulated(self):  # type: (Candidate) -> Candidate
+        """Populate a new Candidate instance with Galaxy signatures.
+        :raises AssertionError: If the supplied candidate is not sourced from a Galaxy-like index.
```
Comment (s-hertel) [inline lib/ansible/galaxy/dependency_resolution/dataclasses.py]:
```suggestion
        :raises AnsibleAssertionError: If the supplied candidate is not sourced from a Galaxy-like index.
```
